File: Android/scenes/dreamFlow/src/main/java/io/agora/scene/dreamFlow/service/DreamFlowService.kt
package io.agora.scene.dreamFlow.service

import android.util.Log
import com.bumptech.glide.load.HttpException
import io.agora.scene.base.BuildConfig
import io.agora.scene.base.EntLogger
import io.agora.scene.dreamFlow.DreamFlowLogger
import kotlinx.coroutines.*
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.logging.HttpLoggingInterceptor
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException

class DreamFlowService constructor(
    private val region: String,
    private val appId: String,
    private val channelName: String,
    private val inUid: Int,
) {

    companion object {
        val genaiUid: Int = 999
    }

    private val tag = "DreamFlowServiceAPI"

    var isEffectOn: Boolean = false

    enum class ServerType {
        Server1,
        Server2,
        Server3,
        Server4;

        fun host(): String {
            return when(this) {
                Server1 -> "http://175.121.93.70:40743"
                Server2 -> "http://175.121.93.70:50249"
                Server3 -> "http://104.15.30.249:49327"
                Server4 -> "http://66.114.112.70:55587"
            }
        }

        fun title(): String {
            return when(this) {
                Server1 -> "Server1"
                Server2 -> "Server2"
                Server3 -> "Server3"
                Server4 -> "Server4"
            }
        }
    }

    enum class Style {
        Toonyou,
        Miyazaki,
        Sexytoon,
        Clay,
        Fantasy;

        fun serverString(): String {
            return when(this) {
                Toonyou -> "toonyou"
                Miyazaki -> "Miyazaki"
                Sexytoon -> "sexytoon"
                Clay -> "Clay"
                Fantasy -> "Fantasy"
            }
        }
    }

    data class SettingBean constructor(
        var isFaceModeOn: Boolean = false,
        var strength: Float = 0.2f,
        var superFrame: Int = 1,
        var style: Style? = Style.Toonyou,
        var prompt: String = ""
    )

    enum class ServiceStatus(val value: String) {
        STARTING("starting"),
        START_SUCCESS("start_success"),
        STOPPED("stopped");

        companion object {
            private val map = values().associateBy { it.value }

            fun fromString(value: String): ServiceStatus {
                return map[value] ?: STOPPED
            }
        }
    }

    var server: ServerType = ServerType.Server2

    var selectedPreset: Int = 0

    var currentSetting: SettingBean? = null

    var status: ServiceStatus = ServiceStatus.STOPPED
        private set

    private var workerId: String = "b3a21856-88a8-4523-b553-cfaf14af5784"

    private val scope = CoroutineScope(Job() + Dispatchers.Main)
    private val okHttpClient by lazy {
        val builder = OkHttpClient.Builder()
        if (BuildConfig.DEBUG) {
            builder.addInterceptor(HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BODY))
        }
        builder.build()
    }

    private var listener: IDreamFlowStateListener? = null
    fun setListener(l: IDreamFlowStateListener) {
        listener = l
    }

    fun clean() {
        endStateTimer()
        listener = null
    }

    private var job: Job? = null
    private fun startStateTimer() {
        job?.cancel()
        job = scope.launch(Dispatchers.Main) {
            repeat(100) {
                delay(3000)
                try {
                    getStatus()
                } catch (e: Exception) {
                    listener?.onOccurError(e)
                }
            }
        }
    }

    private fun endStateTimer() {
        job?.cancel()
        job = null
    }

    fun updateStarted() {
        updateProgress(100)
        updateStatus(ServiceStatus.START_SUCCESS)
    }

    private fun updateStatus(s: ServiceStatus) {
        if (status != s) {
            status = s
            when (status) {
                ServiceStatus.STARTING -> {
                    startStateTimer()
                }
                ServiceStatus.START_SUCCESS,
                ServiceStatus.STOPPED -> {
                    endStateTimer()
                }
            }
            scope.launch(Dispatchers.Main) {
                listener?.onStatusChanged(status)
            }
        }
    }

    private fun updateProgress(p: Int) {
        if (status == ServiceStatus.STARTING) {
            listener?.onLoadingProgressChanged(p)
        }
    }

    fun save(effectOn: Boolean,
             settingBean: SettingBean,
             success: () -> Unit,
             failure: ((Exception?) -> Unit)? = null) {
        scope.launch(Dispatchers.Main) {
            isEffectOn = effectOn
            try {
                if (isEffectOn) {
                    // turn effect on
                    if (status == ServiceStatus.STARTING ||
                        status == ServiceStatus.START_SUCCESS) {
                        currentSetting = settingBean
                        update(settingBean)
                        success.invoke()
                    } else {
                        currentSetting = settingBean
                        val code = create(settingBean)
                        success.invoke()
                        if (code == 0) {
                            updateStatus(ServiceStatus.STARTING)
                        } else {
                            updateStatus(ServiceStatus.START_SUCCESS)
                        }
                    }
                } else {
                    // turn effect off
                    delete(false)
                    currentSetting = settingBean
                    success.invoke()
                    updateStatus(ServiceStatus.STOPPED)
//                    if (status == ServiceStatus.STARTING ||
//                        status == ServiceStatus.STARTED) {
//                        delete(false)
//                        currentSetting = settingBean
//                        success.invoke()
//                        updateStatus(ServiceStatus.IDLE)
//                    } else {
//                        currentSetting = settingBean
//                        success.invoke()
//                        updateStatus(ServiceStatus.IDLE)
//                    }
                }
            } catch (e: Exception) {
                failure?.invoke(e)
            }
        }
    }

    private fun create(
        settingBean: SettingBean,
        success: () -> Unit,
        failure: ((Exception?) -> Unit)? = null
    ) {
        scope.launch(Dispatchers.Main) {
            try {
                create(settingBean)
                success.invoke()
            } catch (e: Exception) {
                failure?.invoke(e)
            }
        }
    }

    fun delete(
        success: () -> Unit,
        failure: ((Exception?) -> Unit)? = null
    ) {
        scope.launch(Dispatchers.Main) {
            try {
                delete(true)
                success.invoke()
            } catch (e: Exception) {
                failure?.invoke(e)
            }
        }
    }

    private suspend fun create(settingBean: SettingBean) = withContext(Dispatchers.IO) {
        val postBody = JSONObject().apply {
            put("name", "agoralive")
            put("style", settingBean.style?.serverString())
            put("strength", settingBean.strength)
            put("faceMode", settingBean.isFaceModeOn)
            put("superFrameFactor", settingBean.superFrame)
            put("rtcConfigure", JSONObject().apply {
                put("userids", JSONArray().apply {
                    put(JSONObject().apply {
                        put("inUid", inUid)
                        put("inChannelName", channelName)
                        put("genaiUid", genaiUid)
                        put("genaiToken", "")
                        put("genaiVideoWidth", 720)
                        put("genaiVideoHeight", 1280)
                        put("prompt", settingBean.prompt)
                    })
                })
            })
        }
        DreamFlowLogger.d(tag, "create, $postBody")
        val request = Request.Builder().url("${server.host()}/$region/v1/projects/$appId/stylize").
        addHeader("Content-Type", "application/json").post(postBody.toString().toRequestBody()).build()
        val execute = okHttpClient.newCall(request).execute()
        if (execute.isSuccessful) {
            val body = execute.body ?: throw RuntimeException("error: ${execute.code} message: ${execute.message}")
            Log.d(tag, "create result body: $body")
            val bodyJobj = JSONObject(body.string())
            Log.d(tag, "create result body obj: $bodyJobj")
            if (bodyJobj["code"] != 0 && bodyJobj["code"] != 1) {
                throw RuntimeException("error: ${execute.code} message: ${execute.message}")
            } else {
                val data = bodyJobj["data"] as JSONObject
                workerId = data["id"] as String
                Log.d(tag, "create result succeed: $data workerId: $workerId")
                bodyJobj["code"]
            }
        } else {
            val body = execute.body ?: throw RuntimeException("error: ${execute.code} message: ${execute.message}")
            val obj = JSONObject(body.string())
            val code = obj["code"] as Int
            val msg = obj["message"] as String
            throw RuntimeException("error: $code message: $msg")
        }
    }

    private suspend fun delete(admin: Boolean) = withContext(Dispatchers.IO) {
        DreamFlowLogger.d(tag, "delete, $workerId")
        val postBody = JSONObject().apply {
            put("admin", admin)
        }
        val request = Request.Builder().url("${server.host()}/$region/v1/projects/$appId/stylize/$workerId").
        addHeader("Content-Type", "application/json").delete(postBody.toString().toRequestBody()).build()
        val execute = okHttpClient.newCall(request).execute()
        if (execute.isSuccessful) {
            val body = execute.body ?: throw RuntimeException("error: ${execute.code} message: ${execute.message}")
            val bodyJobj = JSONObject(body.string())
            Log.d(tag, "delete result body obj: $bodyJobj")
            val code = bodyJobj["code"]
            if (code != 0) {
                throw RuntimeException("error: ${execute.code} message: ${execute.message}")
            } else {
                // update state
                workerId = ""
                code
            }
        } else {
            throw RuntimeException("error: ${execute.code} message: ${execute.message}")
        }
    }

    private suspend fun update(settingBean: SettingBean) = withContext(Dispatchers.IO) {
        val postBody = JSONObject().apply {
            put("name", "agoralive")
            put("style", settingBean.style?.serverString())
            put("strength", settingBean.strength)
            put("faceMode", settingBean.isFaceModeOn)
            put("superFrameFactor", settingBean.superFrame)
            put("rtcConfigure", JSONObject().apply {
                put("userids", JSONArray().apply {
                    put(JSONObject().apply {
                        put("inUid", inUid)
                        put("inChannelName", channelName)
                        put("genaiUid", genaiUid)
                        put("genaiToken", "")
                        put("genaiVideoWidth", 720)
                        put("genaiVideoHeight", 1280)
                        put("prompt", settingBean.prompt)
                    })
                })
            })
        }
        DreamFlowLogger.d(tag, "update, worker: $workerId, body: $postBody")
        val request = Request.Builder().url("${server.host()}/$region/v1/projects/$appId/stylize/$workerId").
        addHeader("Content-Type", "application/json").patch(postBody.toString().toRequestBody()).build()
        val execute = okHttpClient.newCall(request).execute()
        if (execute.isSuccessful) {
            val body = execute.body ?: throw RuntimeException("error: ${execute.code} message: ${execute.message}")
            val bodyJobj = JSONObject(body.string())
            Log.d(tag, "update result body obj: $bodyJobj")
            if (bodyJobj["code"] != 0) {
                throw RuntimeException("error: ${execute.code} message: ${execute.message}")
            } else {
                // do noting
            }
        } else {
            throw RuntimeException("error: ${execute.code} message: ${execute.message}")
        }
    }

    private suspend fun getStatus() = withContext(Dispatchers.IO) {
        DreamFlowLogger.d(tag, "getStatus, worker: $workerId")
        val request = Request.Builder().url("${server.host()}/$region/v1/projects/$appId/stylize/$workerId").
        addHeader("Content-Type", "application/json").get().build()
        val execute = okHttpClient.newCall(request).execute()
        if (execute.isSuccessful) {
            val body = execute.body ?: throw RuntimeException("error: ${execute.code} message: ${execute.message}")
            val bodyJobj = JSONObject(body.string())
            val code = bodyJobj["code"]
            Log.d(tag, "status result body obj: $bodyJobj")
            if (code == 0) {
                val data = bodyJobj["data"] as JSONObject
                val workerInfo = data.optJSONObject("workerInfo")
                val stylize_ai = workerInfo?.optJSONObject("stylize_ai")
                val stateStr = stylize_ai?.optString("state")
                if (stateStr != null) {
                    val status = ServiceStatus.fromString(stateStr)
                    updateStatus(status)
                }
                code
            } else if (code == 10003) {
                updateStatus(ServiceStatus.STOPPED)
                val message = bodyJobj["message"]
                throw RuntimeException("error: $code $message")
            } else {
                throw RuntimeException("error: ${execute.code} message: ${execute.message}")
            }
        } else {
            throw RuntimeException("error: ${execute.code} message: ${execute.message}")
        }
    }
}

File: iOS/AgoraEntScenarios/Scenes/DreamFlow/DreamFlow/Classes/Service/DreamFlowService.swift
//
//  DreamFlowService.swift
//  DreamFlow
//
//  Created by qinhui on 2024/9/18.
//

import Foundation
import AgoraCommon

enum DreamFlowWorkState {
    case unload
    case running
    case initialize
    case failed
}

class DreamFlowService {
    let region = "cn"
    let inUid = 0
//    let inRole = 0
    let genaiUid = 999
    var workState: DreamFlowWorkState = .unload
    var responseModel: DreamFlowResponseModel?
    var currentConfig: DFStylizedSettingConfig?
    var server: String = "http://175.121.93.70:50249"
    
    func creatWork(channelName: String, stylizedConfig:DFStylizedSettingConfig, completion: @escaping ((Error?, DreamFlowWorkState?) -> Void)) {
        currentConfig = stylizedConfig
        
        var requestModel = DreamFlowCreatWorkModel(region: region, appId: AppContext.shared.appId)
        let rtcConfigure = DreamFlowUidConfig()
        let config = DreamFlowRtcConfig()
        config.inUid = Int(UserInfo.userId) ?? 0
//        config.inToken = AppContext.shared.appId
        config.inChannelName = channelName
        config.genaiToken = stylizedConfig.aiToken
//        config.inRole = inRole
        config.genaiUid = genaiUid
        config.prompt = stylizedConfig.prompt
        config.genaiVideoWidth = stylizedConfig.videoWidth
        config.genaiVideoHeight = stylizedConfig.videHeight
        
        rtcConfigure.userids = [config]
        requestModel.name = "agoralive"
        requestModel.rtcConfigure = rtcConfigure
        requestModel.style = stylizedConfig.style
        requestModel.strength = stylizedConfig.strength
        requestModel.faceMode = stylizedConfig.face_mode
        requestModel.superFrameFactor = stylizedConfig.superFrameFactor
        requestModel.host = server
        requestModel.request { error, res in
            if error != nil {
                completion(error, .unload)
                ShowLogger.error("create worker error :\(error?.localizedDescription)")
                return
            }
            
            if let response = VLResponseData.yy_model(withJSON: res), let responseData = response.data {
                guard let model = DreamFlowResponseModel.yy_model(withJSON: responseData) else {
                    ShowLogger.error("create worker analysis json error")
                    completion(error, nil)
                    return
                }
                
                self.responseModel = model
                
                guard let stat = model.stat else {
                    self.workState = .failed
                    ShowLogger.error("work start failed")
                    completion(nil, self.workState)
                    return
                }
                
                if stat == "starting" {
                    self.workState = .initialize
                } else if stat == "running" {
                    self.workState = .running
                } else {
                    self.workState = .failed
                }
                
                ShowLogger.info("work create success")
                completion(nil, self.workState)
                return
            }
            
            ShowLogger.error("Analysis json error")
            completion(error, nil)
        }
    }
    
    func deleteWorker(workerId: String, completion: @escaping ((Error?, Any?) -> Void)) {
        let requestModel = DreamFlowDeleteWorkModel(region: region, appId: AppContext.shared.appId, workerId: workerId)
        requestModel.host = server
        requestModel.request { error, res in
            if error != nil {
                ShowLogger.error("delete worker: \(workerId) failed error: \(error?.localizedDescription)")
                print("delete worker: \(workerId) failed error: \(error?.localizedDescription)")
            } else {
                ShowLogger.info("delete worker success, work Id: \(workerId)")
                self.workState = .unload
            }
            
            completion(error, res)
        }
    }
    
    func updateWorker(workerId: String, stylizedConfig:DFStylizedSettingConfig, completion: @escaping ((Error?, Any?) -> Void)) {
        let requestModel = DreamFlowUpdateWorkModel(region: region, appId: AppContext.shared.appId, workerId: workerId)
        requestModel.prompt = stylizedConfig.prompt
        requestModel.strength = stylizedConfig.strength
        requestModel.superFrameFactor = stylizedConfig.superFrameFactor
        requestModel.style = stylizedConfig.style
        requestModel.host = server
        requestModel.request { err, res in
            if let error = err {
                ShowLogger.error("request update worker failed: \(error.localizedDescription)")
                completion(err, nil)
                return
            }
            
            ShowLogger.info("update worker success")
            completion(nil, res)
        }
    }
    
    func deleteAllWorker(completion: @escaping ((Error?, Any?) -> Void)) {
        let requestModel = DreamFlowDeleteWorkModel(region: region, appId: AppContext.shared.appId, workerId: "b3a21856-88a8-4523-b553-cfaf14af5784")
        requestModel.host = server
        requestModel.request { error, res in
            if error != nil {
                ShowLogger.error("delete worker failed error: \(error?.localizedDescription)")
                print("delete all worker failed error: \(error?.localizedDescription)")
            } else {
                ShowLogger.info("delete all worker success")
            }
            
            completion(error, res)
        }
    }
    
    func queryWorker(workerId: String, completion: @escaping ((Error?, String?) -> Void)) {
        let requestModel = DreamFlowQueryWorkModel(region: region, appId: AppContext.shared.appId, workerId: workerId)
        requestModel.host = server
        requestModel.request { error, res in
            if let err = error {
                ShowLogger.error("request query worker failed error: \(err.localizedDescription)")
                completion(err, nil)
                return
            }
            
            if let response = VLResponseData.yy_model(withJSON: res),
                let responseData = response.data as? [String: Any],
                let workerInfo = responseData["workerInfo"] as? [String: Any],
                let stylizeAi = workerInfo["stylize_ai"] as? [String: Any],
                let state = stylizeAi["state"] as? String {
                ShowLogger.info("query worker success")
                completion(nil, state)
            } else {
                ShowLogger.error("query worker json failed error: \(error?.localizedDescription)")

                completion(error, nil)
            }
        }
    }
}


